runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/data/BlockTransferConnectionQueue.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.nemo.runtime.executor.data;
+
+import edu.snu.nemo.conf.JobConf;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A class to restrict parallel connection per runtime edge.
+ */
+public final class BlockTransferConnectionQueue {
+  private static final Object FUTURE_OBJECT = new Object();
+  private final Map<String, Integer> runtimeEdgeIdToNumOutstandingConnections = new ConcurrentHashMap<>();
```

Comment:
Does `OutstandingConnections` have special meaning? 
If not, how about simple `Connections` instead, as another variable uses simple pending`Connections`?

Also, is `runtimeEdgeIdToNumOutstandingConnections` really needed, as the number can be inferred from `runtimeEdgeIdToPendingConnections.get(runtimeEdgeId).size()`? 
It seems that by using these two variables, two points of modification are made on handling connections. For example, in the following code in `connectionFinished`, `currentOutstandingConnections - 1` can be minus if not carefully handled!

> runtimeEdgeIdToNumOutstandingConnections.put(runtimeEdgeId, currentOutstandingConnections - 1);